### PR TITLE
Prototype

### PR DESCRIPTION
**Pull request by richfitz**
Prototype implementation. No real tests at all! Minimal CI!

But this needs so much that will be awful to get online atm

### CODE REVIEW

**Comment by richfitz** [comment]
```
Warning messages:
1: In file.rename(private$workdir, path_pack) :
  cannot rename file '/orderly/draft/rtm_inference_pmcmc_spim_fits/20201103-151437-ed3a67ae' to '/tmp/Rtmp0laDfe/file27ad324d0584/20201103-151437-ed3a67ae/pack', reason 'Invalid cross-device link'
```

File: DESCRIPTION [hunk 1]
```diff
@@ -6,15 +6,30 @@ Authors@R: c(person("Rich", "FitzJohn", role = c("aut", "cre"),
...
-  didehpc
+    context,
+    didehpc,
+    provisionr,
+    testthat
+Remotes:
+    mrc-ide/context,
+    mrc-ide/didehpc,
+    mrc-ide/provisionr,
+    reside-ic/spud,
```
Comment (weshinsley) [inline DESCRIPTION:34]: Should this be ncov-ic/spud ?

Comment (weshinsley) [inline DESCRIPTION:34]: Should this be ncov-ic/spud

File: R/cluster.R [hunk 2]
```diff
@@ -0,0 +1,19 @@
+prepare_cluster <- function(packages, initialise = TRUE) {
+  ## Ed: Interactively on Windows, I need to do a web_login this
```
Comment (weshinsley) [inline R/cluster.R:2]: My grammar failure - remove the "this"

File: R/main.R [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    mode <- "list"
+    target <- roundtable_list
+    args <- list()
+  } else if (dat$prepare) {
+    mode <- "prepare"
+    target <- roundtable_rtm_prepare
+    args <- dat[c("short_run", "test_run", "date", "chains", "name")]
+    args$chains <- as.integer(args$chains)
+  } else if (dat$run) {
+    mode <- "status"
```
Comment (weshinsley) [inline R/main.R:34]: Should "status" be "run" ?
(That said - mode is not used in the do.call on 16 - is it needed at all?)

File: R/main.R [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+  if (dat$list) {
+    mode <- "list"
+    target <- roundtable_list
+    args <- list()
+  } else if (dat$prepare) {
+    mode <- "prepare"
+    target <- roundtable_rtm_prepare
+    args <- dat[c("short_run", "test_run", "date", "chains", "name")]
+    args$chains <- as.integer(args$chains)
```
Comment (weshinsley) [inline R/main.R:32]: What happens if we omit chains? (as it's an [option])

File: NAMESPACE [hunk 5]
```diff
@@ -1 +1,7 @@
 # Generated by roxygen2: do not edit by hand
+
+export(roundtable_debug)
+export(roundtable_import)
+export(roundtable_list)
+export(roundtable_rtm_prepare)
+export(roundtable_status)
```
Comment (weshinsley) [inline NAMESPACE:7]: Do we need a re-document to add rountable_run ?

File: R/roundtable.R [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+##' @title Run roundtable job on cluster
+##'
+##' @param key The job key
+##'
+##' @param test_job Send a test job to the cluster to ensure that
+##'   everything works.
+##'
+##' @param rerun Re-queue/rerun jobs even if they have been
+##'   completed. Rewrites the `id` on sharepoint.
+##'
```
Comment (weshinsley) [inline R/roundtable.R:15]: Need to include `rerun` in the docopt / parse_args section?

File: R/main.R [hunk 7]
```diff
@@ -0,0 +1,61 @@
+usage <- "Usage:
+  roundtable list
+  roundtable prepare [options] <name>
+  roundtable run [--test-job] <key>
```
Comment (weshinsley) [inline R/main.R:4]: Do we also need [--rerun] here?

File: R/main.R [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    args <- list()
+  } else if (dat$prepare) {
+    mode <- "prepare"
+    target <- roundtable_rtm_prepare
+    args <- dat[c("short_run", "test_run", "date", "chains", "name")]
+    args$chains <- as.integer(args$chains)
+  } else if (dat$run) {
+    mode <- "status"
+    target <- roundtable_run
+    args <- dat[c("test_job", "key")]
```
Comment (weshinsley) [inline R/main.R:36]: Possibly also "rerun"

File: R/roundtable.R [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+##' @author Richard Fitzjohn
+roundtable_run <- function(key, test_job = FALSE, rerun = FALSE) {
+  folder <- sharepoint_folder()
+  incoming <- roundtable_incoming_download(key, folder)
+
+  if (incoming$status$finished) {
```
Comment (weshinsley) [inline R/roundtable.R:21]: `&& !rerun` ?

File: R/roundtable.R [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+  obj <- prepare_cluster(names(incoming$packages), initialise = TRUE)
+  if (test_job) {
+    t <- obj$enqueue(utils::packageVersion("sircovid"))
+    res <- t$wait(timeout = 100)
+    if (t$status() == "ERROR") {
+      stop("Error running test job: ", res$message)
+    }
+  }
+
+  if (incoming$status$running && !rerun) {
```
Comment (weshinsley) [inline R/roundtable.R:34]: Is it safe to allow rerun if the job is still actually running?

File: R/roundtable.R [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+##'
+##' @title Run roundtable job on cluster
+##'
+##' @param key The job key
+##'
+##' @param test_job Send a test job to the cluster to ensure that
+##'   everything works.
+##'
+##' @param rerun Re-queue/rerun jobs even if they have been
+##'   completed. Rewrites the `id` on sharepoint.
```
Comment (weshinsley) [inline R/roundtable.R:14]: I think the code also allows rerun if the job is still running - ? - see below

File: R/main.R [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+  dir.create(path, FALSE, TRUE)
+  rscript <- "#!/usr/bin/env Rscript"
+  code <- c(rscript,
+            "options(pillar.min_chars = Inf)",
+            "roundtable:::main()")
+  path_bin <- file.path(path, "roundtable")
+  writeLines(code, path_bin)
+  Sys.chmod(path_bin, "755")
+  invisible(path_bin)
+}
```
Comment (weshinsley) [inline R/main.R:61]: Don't think write_script is used any more?

**Pull request closed without merging**